perceval/simulators/noisy_sampling_simulator.py

```diff
@@ -321,39 +431,59 @@ def samples(self,
         if not check_heralds_detectors(self._heralds, self._detectors):
             return {"results": BSSamples(), "physical_perf": 1, "logical_perf": 0}
 
-        zpp, max_p = self._check_input_svd(svd)
-
         # Choose a consistent samples limit
         prepare_samples = max_samples
         if max_shots is not None:
             prepare_samples = min(max_samples, max_shots)
         if prepare_samples == 0:
             return {"results": BSSamples(), "physical_perf": 1, "logical_perf": 1}
 
+        source_defined = isinstance(svd, tuple)
+
         # The best case scenario is a perfect sampling => use the "highway" code
         det_type = get_detection_type(self._detectors)
-        if not self._heralds and not self._postselect.has_condition and len(svd) == 1 and det_type == DetectionType.PNR:
-            only_input = next(iter(svd))[0]
+        one_input = len(svd) == 1 or (source_defined and svd[0].is_perfect())
+        if not self._heralds and not self._postselect.has_condition and one_input and det_type == DetectionType.PNR:
+            only_input = svd[1] if source_defined else next(iter(svd))[0]
             if not only_input.has_annotations:
                 get_logger().debug("Perfect sampling: use the fast '_perfect_samples_no_selection' call",
                                    channel.general)
                 return self._perfect_sampling_no_selection(only_input, prepare_samples, progress_callback)
 
-        new_input, pre_physical_perf = self._preprocess_input_state(svd, max_p, prepare_samples)
-
-        # Prepare pools of pre-computed samples
         provider = SamplesProvider(self._backend)
         provider.sleep_between_batches = self.sleep_between_batches
-        provider.prepare(new_input, prepare_samples, progress_callback)
 
-        res = self._noisy_sampling(new_input, provider, max_samples, max_shots, det_type, progress_callback)
+        if source_defined:
```

Comment:
Would it be possible to split this long methods into steps, to make it clearer. This if / else is hard to understand here in the middle of the code.